sdk/python/kfp/dsl/_container_op.py

```diff
@@ -12,309 +12,940 @@
-          the value of a PipelineParam is saved to its corresponding local file. It's
-          one way for outside world to receive outputs of the container.
-      is_exit_handler: Whether it is used as an exit handler.
-    """
-
-    if not _pipeline.Pipeline.get_default_pipeline():
-      raise ValueError('Default pipeline not defined.')
-
-    valid_name_regex = r'^[A-Za-z][A-Za-z0-9\s_-]*$'
-    if not re.match(valid_name_regex, name):
-      raise ValueError('Only letters, numbers, spaces, "_", and "-"  are allowed in name. Must begin with letter: %s' % (name))
-
-    self.human_name = name
-    self.name = _pipeline.Pipeline.get_default_pipeline().add_op(self, is_exit_handler)
-    self.image = image
-    self.command = command
-    self.arguments = arguments
-    self.is_exit_handler = is_exit_handler
-    self.resource_limits = {}
-    self.resource_requests = {}
-    self.node_selector = {}
-    self.volumes = []
-    self.volume_mounts = []
-    self.env_variables = []
-    self.pod_annotations = {}
-    self.pod_labels = {}
-    self.num_retries = 0
-    self._metadata = None
-
-    self.argument_inputs = _extract_pipelineparams([str(arg) for arg in (command or []) + (arguments or [])])
-
-    self.file_outputs = file_outputs
-    self.dependent_op_names = []
-
-    self.inputs = []
-    if self.argument_inputs:
-      self.inputs += self.argument_inputs
-
-    self.outputs = {}
-    if self.file_outputs:
-      self.outputs = {name: _pipeline_param.PipelineParam(name, op_name=self.name)
-                      for name in file_outputs.keys()}
-
-    self.output=None
-    if len(self.outputs) == 1:
-      self.output = list(self.outputs.values())[0]
-
-  def apply(self, mod_func):
-    """Applies a modifier function to self. The function should return the passed object.
-    This is needed to chain "extention methods" to this class.
-
-    Example:
-      from kfp.gcp import use_gcp_secret
-      task = (
-        train_op(...)
-          .set_memory_request('1GB')
-          .apply(use_gcp_secret('user-gcp-sa'))
-          .set_memory_limit('2GB')
-      )
-    """
-    return mod_func(self)
-
-  def after(self, op):
-    """Specify explicit dependency on another op."""
-    self.dependent_op_names.append(op.name)
-    return self
-
-  def _validate_memory_string(self, memory_string):
-    """Validate a given string is valid for memory request or limit."""
 
-    if re.match(r'^[0-9]+(E|Ei|P|Pi|T|Ti|G|Gi|M|Mi|K|Ki){0,1}$', memory_string) is None:
-      raise ValueError('Invalid memory string. Should be an integer, or integer followed '
-                       'by one of "E|Ei|P|Pi|T|Ti|G|Gi|M|Mi|K|Ki"')
+# generics
+T = TypeVar('T')
+# type alias: either a string or a list of string
+StringOrStringList = Union[str, List[str]]
 
-  def _validate_cpu_string(self, cpu_string):
-    "Validate a given string is valid for cpu request or limit."
 
-    if re.match(r'^[0-9]+m$', cpu_string) is not None:
-      return
+# util functions
```

Comment:
Let's move the general utility functions into a separate module.

Reply:
This is more container_op related but I think it would be helpful to define a decorator that helps for the future deprecation warnings.